File: algoritmos/mergesort.js
// MergeSort para Arrays

function splitArray(array) {
    const middle = Math.floor(array.length / 2);
    const left = array.slice(0, middle);
    const right = array.slice(middle);
    
    return { left, right };
}

function mergeArrays(left, right) {
    const result = [];
    let leftIndex = 0;
    let rightIndex = 0;
    
    while (leftIndex < left.length && rightIndex < right.length) {
        if (left[leftIndex] < right[rightIndex]) {
            result.push(left[leftIndex]);
            leftIndex++;
        } else {
            result.push(right[rightIndex]);
            rightIndex++;
        }
    }
    
    const remainingLeft = left.slice(leftIndex);
    const remainingRight = right.slice(rightIndex);
    
    return result.concat(remainingLeft).concat(remainingRight);
}

function mergeSort(array) {
    console.log(array);

    if (array.length <= 1) {
        return array;
    }
    
    const { left, right } = splitArray(array);

    console.log("Left:", left);
    console.log("Right:", right);
    
    const leftSorted = mergeSort(left);
    const rightSorted = mergeSort(right);
    
    return mergeArrays(leftSorted, rightSorted);
}

const array = [4, 2, 1, 3, 5, 6];

console.log(mergeSort(array));

File: algoritmos/mergesort.go
package main

import (
	"fmt"
)

func main() {
	array := []int{4, 2, 1, 3, 5}

	fmt.Println(mergeSort(array))
}

func splitArray(array []int) ([]int, []int) {
	middle := len(array) / 2
	left := array[:middle]
	right := array[middle:]

	return left, right
}

func mergeArrays(left, right []int) []int {
	result := make([]int, 0, len(left)+len(right))
	leftIndex := 0
	rightIndex := 0

	for leftIndex < len(left) && rightIndex < len(right) {
		if left[leftIndex] < right[rightIndex] {
			result = append(result, left[leftIndex])
			leftIndex++
		} else {
			result = append(result, right[rightIndex])
			rightIndex++
		}
	}

	result = append(result, left[leftIndex:]...)
	result = append(result, right[rightIndex:]...)

	return result
}

func mergeSort(array []int) []int {
	fmt.Println("Array: ", array)

	if len(array) <= 1 {
		return array
	}

	left, right := splitArray(array)

	fmt.Println("Left:", left)
	fmt.Println("Right:", right)

	leftSorted := mergeSort(left)
	rightSorted := mergeSort(right)

	fmt.Println("leftSorted:", leftSorted)
	fmt.Println("rightSorted:", rightSorted)

	merge := mergeArrays(leftSorted, rightSorted)
	fmt.Println("merge", merge)

	return merge
}
